File: src/AutoStart.ts
import { writeFileSync, unlinkSync, existsSync, mkdirSync } from "node:fs"
import { homedir, platform } from "node:os"
import path from "node:path"
import { execSync } from "node:child_process"

/**
 * Lightweight cross-platform autostart helper.
 * - Linux: creates a .desktop file in ~/.config/autostart
 * - macOS: creates a LaunchAgent plist in ~/Library/LaunchAgents
 * - Windows: sets HKCU Run registry value
 */
export class AutoStart {
    private static appName = "lyrics-status"

    public static enable(enable: boolean): void {
        try {
            const plt = platform()
            if (plt === "linux") return this.enableLinux(enable)
            if (plt === "darwin") return this.enableMac(enable)
            if (plt === "win32") return this.enableWindows(enable)
        } catch (e) {
            // best-effort; do not crash the app
            // eslint-disable-next-line no-console
            console.error("AutoStart error:", (e as Error).message)
        }
    }

    private static nodeCommand(): string {
        // Return a quoted absolute command so autostart works regardless of cwd.
        const node = process.execPath
        const script = process.argv[1] ? path.resolve(process.argv[1]) : ""
        if (!script) return `"${node}"`
        return `"${node}" "${script}"`
    }

    private static enableLinux(enable: boolean): void {
        const configDir = path.join(homedir(), ".config", "autostart")
        if (!existsSync(configDir)) mkdirSync(configDir, { recursive: true })

        const desktopPath = path.join(configDir, `${this.appName}.desktop`)

        if (!enable) {
            if (existsSync(desktopPath)) unlinkSync(desktopPath)
            return
        }

        const exec = this.nodeCommand()
        const content = [
            "[Desktop Entry]",
            `Type=Application`,
            `Name=${this.appName}`,
            `Exec=${exec}`,
            `X-GNOME-Autostart-enabled=true`,
        ].join("\n")

        writeFileSync(desktopPath, content, { encoding: "utf8" })
    }

    private static enableMac(enable: boolean): void {
        const agents = path.join(homedir(), "Library", "LaunchAgents")
        if (!existsSync(agents)) mkdirSync(agents, { recursive: true })

        const label = `com.${this.appName}.autostart`
        const plistPath = path.join(agents, `${label}.plist`)

        if (!enable) {
            if (existsSync(plistPath)) unlinkSync(plistPath)
            try { execSync(`launchctl unload ${plistPath}`) } catch {}
            return
        }

        const exec = this.nodeCommand()
        const plist = `<?xml version="1.0" encoding="UTF-8"?>\n<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">\n<plist version="1.0">\n<dict>\n  <key>Label</key>\n  <string>${label}</string>\n  <key>ProgramArguments</key>\n  <array>\n    <string>${process.execPath}</string>\n    <string>${process.argv[1] || ""}</string>\n  </array>\n  <key>RunAtLoad</key>\n  <true/>\n</dict>\n</plist>`

        writeFileSync(plistPath, plist, { encoding: "utf8" })
        try { execSync(`launchctl load ${plistPath}`) } catch {}
    }

    private static enableWindows(enable: boolean): void {
        // Use HKCU Run to avoid needing admin rights.
        const name = "LyricsStatus"
        const cmd = this.nodeCommand()

        if (!enable) {
            try {
                execSync(`reg delete HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Run /v ${name} /f`)
            } catch {
                // ignore
            }
            return
        }

        try {
            execSync(`reg add HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Run /v ${name} /t REG_SZ /d "${cmd}" /f`)
        } catch (e) {
            // ignore
        }
    }
}

export default AutoStart

File: src/autostart.rs
//! Best-effort autostart: `.desktop` file on Linux, LaunchAgent on macOS,
//! HKCU Run registry entry on Windows.

use std::fs;
use std::path::PathBuf;
#[cfg(target_os = "windows")]
use std::process::Command;

const APP_NAME: &str = "mewsic";

#[cfg(any(target_os = "linux", target_os = "macos"))]
fn home() -> PathBuf {
    std::env::var_os("HOME").map(PathBuf::from).unwrap_or_default()
}

fn command_for_autostart() -> String {
    // Reuse the current executable + a `run` argument so login launches the
    // engine (web panel enabled, headless-safe).
    match std::env::current_exe() {
        Ok(exe) => format!("\"{}\" run", exe.to_string_lossy()),
        Err(_) => "mewsic run".to_string(),
    }
}

/// Apply the autostart setting. Failures are logged, never fatal.
pub fn apply(enabled: bool) {
    let result = apply_inner(enabled);
    if let Err(e) = result {
        crate::log::write(&format!("autostart error: {e}"));
    }
}

fn apply_inner(enabled: bool) -> Result<(), String> {
    #[cfg(target_os = "linux")]
    {
        linux_autostart(enabled)?;
    }
    #[cfg(target_os = "macos")]
    {
        macos_autostart(enabled)?;
    }
    #[cfg(target_os = "windows")]
    {
        windows_autostart(enabled)?;
    }
    Ok(())
}

#[cfg(target_os = "linux")]
fn linux_autostart(enabled: bool) -> Result<(), String> {
    let dir = home().join(".config").join("autostart");
    let file = dir.join(format!("{APP_NAME}.desktop"));
    if !enabled {
        let _ = fs::remove_file(&file);
        return Ok(());
    }
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let exec = command_for_autostart();
    let content = format!(
        "[Desktop Entry]\nType=Application\nName=mewsic\nExec={exec}\nX-GNOME-Autostart-enabled=true\n"
    );
    fs::write(&file, content).map_err(|e| e.to_string())
}

#[cfg(target_os = "macos")]
fn macos_autostart(enabled: bool) -> Result<(), String> {
    let dir = home().join("Library").join("LaunchAgents");
    let label = "com.mewsic.autostart";
    let file = dir.join(format!("{label}.plist"));
    if !enabled {
        let _ = fs::remove_file(&file);
        return Ok(());
    }
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let exe = std::env::current_exe().map_err(|e| e.to_string())?;
    let content = format!(
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\
         <!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n\
         <plist version=\"1.0\">\n<dict>\n\
         \x20 <key>Label</key><string>{label}</string>\n\
         \x20 <key>ProgramArguments</key>\n\
         \x20 <array>\n\
         \x20 \x20 <string>{}</string>\n\
         \x20 \x20 <string>run</string>\n\
         \x20 </array>\n\
         \x20 <key>RunAtLoad</key><true/>\n\
         </dict>\n</plist>\n",
        exe.to_string_lossy()
    );
    fs::write(&file, content).map_err(|e| e.to_string())
}

#[cfg(target_os = "windows")]
fn windows_autostart(enabled: bool) -> Result<(), String> {
    let key = r"HKCU\Software\Microsoft\Windows\CurrentVersion\Run";
    if !enabled {
        let _ = Command::new("reg")
            .args(["delete", key, "/v", APP_NAME, "/f"])
            .status();
        return Ok(());
    }
    let value = command_for_autostart();
    Command::new("reg")
        .args(["add", key, "/v", APP_NAME, "/t", "REG_SZ", "/d", &value, "/f"])
        .status()
        .map_err(|e| e.to_string())?;
    Ok(())
}
